cpp/src/solver/cd.cuh

```diff
@@ -148,59 +178,67 @@ void cdFit(const raft::handle_t& handle,
 
   raft::copy(residual.data(), labels, n_rows, stream);
 
+  ConvState<math_t> h_convState;
+  rmm::device_uvector<ConvState<math_t>> convStateBuf(1, stream);
+  auto convStateLoc = convStateBuf.data();
+
+  CUBLAS_CHECK(
+    raft::linalg::cublassetpointermode(cublas_handle, CUBLAS_POINTER_MODE_DEVICE, stream));
+  rmm::device_scalar<math_t> cublas_alpha(1.0, stream);
+  rmm::device_scalar<math_t> cublas_beta(0.0, stream);
+
   for (int i = 0; i < epochs; i++) {
+    // RAFT_USING_NVTX_RANGE(stream, "ML::Solver::cdFit::epoch-%d", i);
     if (i > 0 && shuffle) { Solver::shuffle(ri, g); }
 
-    math_t coef_max   = 0.0;
-    math_t d_coef_max = 0.0;
-    math_t coef_prev  = 0.0;
+    CUDA_CHECK(cudaMemsetAsync(convStateLoc, 0, sizeof(ConvState<math_t>), stream));
 
     for (int j = 0; j < n_cols; j++) {
+      // RAFT_USING_NVTX_RANGE(stream, "ML::Solver::cdFit::col-%d", j);
       int ci                = ri[j];
       math_t* coef_loc      = coef + ci;
       math_t* squared_loc   = squared.data() + ci;
       math_t* input_col_loc = input + (ci * n_rows);
 
-      raft::linalg::multiplyScalar(pred.data(), input_col_loc, h_coef[ci], n_rows, stream);
-      raft::linalg::add(residual.data(), residual.data(), pred.data(), n_rows, stream);
-      raft::linalg::gemm(handle,
-                         input_col_loc,
-                         n_rows,
-                         1,
-                         residual.data(),
-                         coef_loc,
-                         1,
-                         1,
-                         CUBLAS_OP_T,
-                         CUBLAS_OP_N,
-                         stream);
-
-      if (l1_ratio > math_t(0.0)) Functions::softThres(coef_loc, coef_loc, alpha, 1, stream);
-
-      raft::linalg::eltwiseDivideCheckZero(coef_loc, coef_loc, squared_loc, 1, stream);
-
-      coef_prev = h_coef[ci];
-      raft::update_host(&(h_coef[ci]), coef_loc, 1, stream);
-      CUDA_CHECK(cudaStreamSynchronize(stream));
-
-      math_t diff = abs(coef_prev - h_coef[ci]);
-
-      if (diff > d_coef_max) d_coef_max = diff;
-
-      if (abs(h_coef[ci]) > coef_max) coef_max = abs(h_coef[ci]);
-
-      raft::linalg::multiplyScalar(pred.data(), input_col_loc, h_coef[ci], n_rows, stream);
-      raft::linalg::subtract(residual.data(), residual.data(), pred.data(), n_rows, stream);
+      raft::copy(&(convStateLoc->coef), coef_loc, 1, stream);
+      CUBLAS_CHECK(raft::linalg::cublasaxpy(
+        cublas_handle, n_rows, coef_loc, input_col_loc, 1, residual.data(), 1, stream));
+
+      raft::linalg::cublasgemv(cublas_handle,
```

Comment:
Yes, I switched to using cublas wrappers directly, because they allow passing `alpha` and `beta` as device pointers. With this change, I'm able to set up `CUBLAS_POINTER_MODE_HOST` once, outside of the loop.

Apparently, repeated calls to `cublassetpointermode` within the inner loop cause a measurable overhead (sorry, I don't have the numbers). Maybe, the reason for that isn't `cublassetpointermode`, but  cublas' `gemv` copying these two scalars between the host and device on every iteration - I don't know. On a related note, I did a small experiment of removing `CUBLAS_CHECK(cublasSetStream(handle, stream))` from all cublas wrappers, and that gave me 5% speedup in the CD solver. Did not worth the trouble this time, but maybe this is something we should think about in the future?..

Reply:
When these wrappers were created in cuml originally, we didn't yet have a cuml/raft handle and so I think this detail was overlooked once we finally added them. Ideally, the cublas internal wrapper functions should be accepting a raft handle because it already contains a cublas handle w/ the main stream from the handle configured on it. 

Though, this detail is yet another reason that we should avoid using the cublas wrapper calls directly. Following my comment [here](https://github.com/rapidsai/cuml/pull/4446#discussion_r768788540), we should add an overload for the gemv that accepts a `rmm::device_scalar` or add a `gemv_device()` function that directly accepts device pointers for alpha and beta (either one is fine since it's temporary) and set the pointer mode only when it doesn't match what is desired.

Reply:
**Update**: I've updated the code to run all operations in the device mode, yet do the switching back to the host mode after each relevant call. Apparently, `cublasSetPointerMode` causes very little overhead. Probably, the performance drops I observed before were due to running some of the operations in the host mode, rather than due to the switching itself.